app/controllers/nodes_controller.rb

```diff
@@ -18,6 +18,42 @@ def show
     @minion = Minion.find(params[:id])
   end
 
+  # rubocop:disable Metrics/AbcSize
+  def assign
+    @minion = Minion.find(params[:node_id])
+
+    unless Minion.roles.keys.include?(params[:role])
+      raise Minion::InvalidRole, "Could not assign role #{params[:role]}"
+    end
+
+    respond_to do |format|
+      if @minion.assign_role(params[:role])
+        res, body = Velum::Salt.orchestrate
+        http_status = res.code == "200" ? :ok : :unprocessable_entity
+        flash[:error] = "Salt failed to orchestrate: #{body.inspect}" unless http_status == :ok
+
+        format.html { redirect_to nodes_path }
+        format.json { head http_status }
+      else
+        format.html do
+          flash[:error] = @minion.errors.full_messages.first
+          redirect_to nodes_path
+        end
+        format.json { render json: @minion.errors, status: :unprocessable_entity }
+      end
+    end
+  rescue Velum::SaltApi::SaltConnectionException,
+         ActiveRecord::RecordNotFound,
+         Minion::InvalidRole => e
```

Comment:
[Most people](https://www.google.gr/search?q=exceptions+as+flow+control&ie=utf-8&oe=utf-8&client=firefox-b-ab&gfe_rd=cr&ei=Hc6qWOfDBKjA8ge7tbPYCA) seem to agree it is bad practice. And to be honest, I also think exceptions should only be raised when something exceptional happens. That is a case which we could not predict and thus we couldn't handle gracefully.